src/style.jl

```diff
@@ -61,4 +70,244 @@ which is the opposite of the image convention.
+    for i in 1:blocklen
+        damges[i] = @views count(in(errinds), blockinds[8 * i - 7:8 * i])
+    end
+    ## sort by damage cost
+    inds = @views sort(1:blocklen, by=i->damges[i], rev=true)[1:destroy]
+    ## recover the codewords
+    @views vcat([blockinds[8 * i - 7:8 * i] for i in inds]...)
+end
+
+"""
+    imageinqrcode(code::QRCode, targetmat::AbstractMatrix, rate::Real=2/3)
+
+Plot the image `targetmat` inside the QR code.
+
+In order to get the best simulation, the error correction level
+is recommended to be `High()`.
+
+The parameter `rate` is the rate of the error correction codewords
+that will be destroyed. The default value is `2/3`, which means that
+`1/2 * 2/3 = 1/3` of the error correction codewords will be destroyed.
+"""
+imageinqrcode(code::QRCode, targtemat::AbstractMatrix; rate::Real=2/3) = imageinqrcode!(copy(code), targtemat; rate=rate)
+
+function imageinqrcode!( code::QRCode
+                       , targetmat::AbstractMatrix
+                       ; rate::Real=2/3)
+    ## check parameters
+    rate ≥ 0 || throw(ArgumentError("rate should be non-negative."))
+    rate ≤ 1 || @warn(ArgumentError("Invalid rate $rate." *
+            " It could risk to destroy the message if rate is bigger that 1."))
+    border, code.border = code.border, 0 # reset border -- compat to indexes rule
+    (imgx, imgy), qrlen = size(targetmat), qrwidth(code)
+    version, eclevel = code.version, code.eclevel
+    qrlen ≥ max(imgx, imgy) || throw(ArgumentError("The image is too large."))
+    
+    ## image position
+    x1, y1 = (qrlen - imgx + 1) >> 1, (qrlen - imgy + 1) >> 1
+    x2, y2 = x1 + imgx - 1, y1 + imgy - 1
+    leftop = (x1 - 1, y1 - 1)
+
+    ## locations of message bits
+    msginds, ecinds = getsegments(version, eclevel)
+    msgecinds = [vcat(inds1, inds2) for (inds1, inds2) in zip(msginds, ecinds)]
+    # number of error correction codewords
+    ncodewords = length(first(ecinds)) >> 3
+    destroy = floor(Int, ncodewords * rate / 2)
+    # indexes inside the image
+    targetval(ind::Int) = targetmat[ind % qrlen - leftop[1], ind ÷ qrlen - leftop[2]]
+    function isvalid(p::Int)
+         x, y = p % qrlen, p ÷ qrlen
+         x1 ≤ x ≤ x2 && y1 ≤ y ≤ y2
+    end
+    validinds = vcat(filter!.(isvalid, msginds)...,
+                     filter!.(isvalid, ecinds)...)
+    validmsgecinds = [vcat(inds1, inds2) for (inds1, inds2) in zip(msginds, ecinds)]
+    ## cost function
+    penalty(newmat) = sum(newmat[validinds] .== targetval.(validinds))
+
+    ## enumerate over masks
+    bestmat, bestpenalty = nothing, Inf
+    for mask in 0:7
+        code.mask = mask
+        newmat = qrcode(code)
+        for (vinds, blockinds) in zip(validmsgecinds, msgecinds) # enumerate over each blcok
+            errinds = filter(x -> newmat[x] != targetval(x), vinds)
+            inds = filter!(isvalid, pickcodewords(errinds, blockinds, destroy))
+            newmat[inds] = targetval.(inds)
```

Comment:
Is it possible to fuse these four loops (`filter`, `pickcodewords`, `filter!`, `targetval.()` into one explicit for loop? I believe that would significantly improve the performance.

Something like

```julia
for (iv, ib) in zip(vinds, blockinds)
    idx = ...
    if isvalid(idx)
        newmat[idx] = targetval(idx)
    end
end
```

Reply:
It might need more designs to fuse four loops. 

Notes for `pickcodewords(errinds, blockinds, destroy)`.

1. `blockinds` can be treated as a list of byte-indexes, like `[a1, a2, ..., a8, b1, b2, ..., b8, ...]`
2. `errinds` locates error indexes.
3. **When we modify one index, we actually change the whole byte.**
4. `pickcodewords` returns bytes that contain errors as many as possible. It contains a sort method, thus cannot accept indexes one by one.

As a result, the first two loops are necessary:
```julia
errinds = filter(x -> newmat[x] != targetval(x), vinds)
inds = pickcodewords(errinds, blockinds, destroy)
```

As for the `isvalid` check, we can replace `msgecinds` by `validmsgecinds` to avoid checking. This could also minimize the computing in `pickcodewords`, since `length(validmsg) <= len(msg)`. But it is much difficult to the byte given only the indexes.